File: week3.2/ransomenote.py
# leetcode problem - 383
# https://leetcode.com/problems/ransom-note/
import collections
def canConstruct(ransomNote, magazine):
    """
    :type ransomNote: str
    :type magazine: str
    :rtype: bool
    """
    # time complexity - O(n)
    # space complexity - O(n)
    # Approach - 1
    # ==============================================
    letterCount = {}
    for letter in magazine:
        if letter in letterCount:
            letterCount[letter] += 1
        else:
            letterCount[letter] = 1

    for letter in ransomNote:
        if letter not in letterCount or letterCount[letter] == 0:
            return False
        else:
            letterCount[letter] -= 1
    return True

    # Approach - 2
    # ==============================================
    # letterCount = collections.Counter(magazine)
    # for letter in ransomNote:
    #     if letter not in letterCount or letterCount[letter] == 0:
    #         return False
    #     else:
    #         letterCount[letter] -= 1
    # return True

print(canConstruct("aa", "aab"))
print(canConstruct("a", "b"))

File: DSA/week3.2/ransomenote.js
// leetcode problem - 383. Ransom Note
// Given two strings ransomNote and magazine, return true if ransomNote can be constructed from magazine and false otherwise.

function canConstruct(ransomNote, magazine) {
    const letterCount = new Map()
    for(let letter of magazine){
        if(letterCount.has(letter)){
            letterCount.set(letter, letterCount.get(letter)+1)
        }else{
            letterCount.set(letter, 1)
        }
    }

    for(let letter of ransomNote){
        if(!letterCount.has(letter)||letterCount.get(letter)===0) return false
        letterCount.set(letter, letterCount.get(letter)-1)
    }
    return true
}
console.log(canConstruct("aa","aab"))
console.log(canConstruct("a","b"))
// time complexity - O(n)
// space complexity - O(n)
